### DISCUSSION TITLE
How to run graph commands on a redis cluster using Jedis.

### DISCUSSION THREAD

**Discussion by gauravguptaiitr** [xetorthio/jedis#3354] [Q&A] [answered]
We have a production redis cluster where we loaded radisgraph module on all the nodes. We have multiple graphs which will be spread across the cluster using the graph name as hash slot. This is the reason why we want to run it in cluster mode.

We have initialized jedis client using a cluster connection provider. We are using the UnifiedJedis interface to call graph queries. 

public ResultSet graphQuery(String name, String query) {
        return (ResultSet)this.executeCommand(this.graphCommandObjects.graphQuery(name, query));
}

This call executes with a CommandObject and not with ClusterCommand object.
public final CommandObject<ResultSet> graphQuery(String name, String query) {
        return new CommandObject((new CommandArguments(GraphCommand.QUERY)).key(name).add(query).add(GraphKeyword.__COMPACT), this.getBuilder(name));
   }

This eventually fails in ClusterconnectionProvider class.

 public Connection getConnection(CommandArguments args) {
        int slot = ((ClusterCommandArguments)args).getCommandHashSlot();
        return slot >= 0 ? this.getConnectionFromSlot(slot) : this.getConnection();
    }

java.lang.ClassCastException: class redis.clients.jedis.CommandArguments cannot be cast to class redis.clients.jedis.ClusterCommandArguments (redis.clients.jedis.CommandArguments and redis.clients.jedis.ClusterCommandArguments

We could not find the graphQuery method which uses a clustercommandobject. 
How could we solve this? Does jedis supports cluster graph queries? Please help with this.
Let us know if any other details are needed.

**Comment by sazzad16** [comment, marked as answer]
@gauravguptaiitr Thank you for reporting. This is a bug. It didn't get caught as we don't have cluster+graph in test base. We'll work on this as soon as we can.

**Reply by gauravguptaiitr** [reply]
Thanks, @sazzad16 for confirmation. We will wait for the fix.

**Reply by sazzad16** [reply]
@gauravguptaiitr I gave a shot fixing it. But sadly we still don't have cluster+graph in our test. Could you please try [latest Jedis snapshot](https://github.com/redis/jedis/blob/master/docs/jedis-maven.md#snapshots) and confirm if this is working or not?

**Reply by gauravguptaiitr** [reply]
Sure @sazzad16. Thanks for fixing this on priority. Let us try this on our setup and confirm.

**Reply by gauravguptaiitr** [reply]
@sazzad16 We have verified the fix on our cluster + graph setup. It is working fine.

**Reply by sazzad16** [reply]
@gauravguptaiitr Great! Thank your for your collaboration.
